osu.Game/Screens/Play/Player.cs

```diff
@@ -483,7 +483,7 @@ private void updatePauseOnFocusLostState()
             }
         }
 
-        private IBeatmap loadPlayableBeatmap(Mod[] gameplayMods)
+        protected virtual IBeatmap LoadPlayableBeatmap(Mod[] gameplayMods, bool useGlobalRuleset = true)
```

Comment:
This boolean param is awkward at best and I don't think it should exist. There are likely better places to set the correct ruleset (hint: `EditorLoader` is probably the correct one).

Reply:
> Still no tests here...

What tests do you mention?

Reply:
> This boolean param is awkward at best and I don't think it should exist. There are likely better places to set the correct ruleset (hint: `EditorLoader` is probably the correct one).

I tried to implement ruleset change in EditorLoader, but there is set "DisallowExternalBeatmapRulesetChanges = true", so ruleset can't be changed from there

Reply:
`DisallowExternalBeatmapRulesetChanges` was *designed* for exactly this purpose. See this existing code:

https://github.com/ppy/osu/blob/77b4bd97f8a73f0c404a56ff408250af3cd671b0/osu.Game/Screens/Edit/EditorLoader.cs#L65-L66

The reason it's not working is an added complication from `SongSelect`, which is caching its own decoupled ruleset bindable over the game global one:

https://github.com/ppy/osu/blob/77b4bd97f8a73f0c404a56ff408250af3cd671b0/osu.Game/Screens/Select/SongSelect.cs#L353-L354

and then attempts to sync it with the global one, which is the root failure here:

https://github.com/ppy/osu/blob/77b4bd97f8a73f0c404a56ff408250af3cd671b0/osu.Game/Screens/Select/SongSelect.cs#L809-L810

This is a bit awkward, but can be bypassed by doing something like this:

```diff
diff --git a/osu.Game/Screens/Edit/EditorLoader.cs b/osu.Game/Screens/Edit/EditorLoader.cs
index 0a2b8437fa..64a77c39f4 100644
--- a/osu.Game/Screens/Edit/EditorLoader.cs
+++ b/osu.Game/Screens/Edit/EditorLoader.cs
@@ -64,6 +64,7 @@ protected override void LoadComplete()
 
             // will be restored via lease, see `DisallowExternalBeatmapRulesetChanges`.
             Mods.Value = Array.Empty<Mod>();
+            Ruleset.Value = Beatmap.Value.BeatmapInfo.Ruleset;
         }
 
         protected virtual Editor CreateEditor() => new Editor(this);
diff --git a/osu.Game/Screens/Select/SongSelect.cs b/osu.Game/Screens/Select/SongSelect.cs
index 8870239485..1755fcc2a8 100644
--- a/osu.Game/Screens/Select/SongSelect.cs
+++ b/osu.Game/Screens/Select/SongSelect.cs
@@ -806,7 +806,17 @@ private void bindBindables()
 
             Ruleset.ValueChanged += r => updateSelectedRuleset(r.NewValue);
 
-            decoupledRuleset.ValueChanged += r => Ruleset.Value = r.NewValue;
+            decoupledRuleset.ValueChanged += r =>
+            {
+                bool wasDisabled = Ruleset.Disabled;
+
+                // a sub-screen may have taken a lease on the decoupled ruleset bindable,
+                // which will disable both the decoupled bindable and the game-global bindable (via the `DisabledChanged` callback below).
+                // to make sure changes sync without crashes, lift the disable for a second to sync, and then restore the old value.
+                Ruleset.Disabled = false;
+                Ruleset.Value = r.NewValue;
+                Ruleset.Disabled = wasDisabled;
+            };
             decoupledRuleset.DisabledChanged += r => Ruleset.Disabled = r;
 
             Beatmap.BindValueChanged(workingBeatmapChanged);
```

which I think is still better than what you have here, but I'm going to wait for further opinions to make sure there's agreement on that.

Reply:
Now things became more clear after your explanation, ty
But thing that important variables are changing in loader screen that main function seems exist only to make smooth transition between screen bothers me a bit
Also this construction with temporary disabling binds feels like pattern violation
In the same time I understand your solution and it seems more suitable for current case than my previous solutions

BTW, the source of issue in convertible maps (this ruleset workaround need only to work with convertible maps that edited in their native mode), so true issue probably lie a bit deeper
But this solutions seems fine, so ty for help

Reply:
Could you please reply what tests do you mean in your first comment?
You mentioned that two times, but I haven't figured out exactly what you mean

I saw test classes with methods, but both my solutions seems minor so I thought that such test methods weren't needed for them

Reply:
My proposed solution is a bit of a hack yes and I'm not denying it, but I'm saying it feels less of a hack than adding the parameter to `LoadPlayableBeatmap` proposed here. If we're loading a beatmap then we should be loading beatmap, no ifs, ands or buts - the beatmap's ruleset should be used rather than some ruleset from an external beatmap.

> Could you please reply what tests do you mean in your first comment?

Ideally I'm expecting to see a test (probably in `TestSceneScreenNavigation`) that asserts that if a converted beatmap is opened in editor, then the ruleset changes to the original beatmap ruleset while editor is open, and then reverts to the previous value when editor is exited. The solution may be "minor" but the point of adding the test is to serve as documentation/paper trail that this was an intended behaviour that should not be broken in the future.

Reply:
So, to finish this pull request I need to create test method for this case (as issue seems solved already)
Correct?